lib/Sema/ConstraintSystem.cpp

```diff
@@ -2749,239 +2897,130 @@ bool ConstraintSystem::diagnoseAmbiguityWithFixes(
-    // Let's apply the solution so the contextual generic types
-    // are available in the system for diagnostics.
-    applySolution(*viableSolutions[0]);
-    solutions.front().Fixes.front()->diagnose(/*asNote*/ false);
-    return true;
+    return diagnosed;
   }
 
-  if (!diagnosable || viableSolutions.size() < 2)
-    return false;
-
-  auto *decl = *distinctChoices.begin();
+  auto *commonCalleeLocator = ambiguousOverload->locator;
+  auto *decl = ambiguousOverload->choices.front().getDecl();
   assert(solverState);
 
   bool diagnosed = true;
   {
     DiagnosticTransaction transaction(getASTContext().Diags);
 
     auto *commonAnchor = commonCalleeLocator->getAnchor();
+    if (auto *callExpr = dyn_cast<CallExpr>(commonAnchor))
+      commonAnchor = callExpr->getDirectCallee();
     auto &DE = getASTContext().Diags;
     auto name = decl->getFullName();
 
-    auto emitGeneralAmbiguityFailure = [&]() {
-      // Three choices here:
-      // 1. If this is a special name avoid printing it because
-      //    printing kind is sufficient;
-      // 2. If all of the labels match, print a full name;
-      // 3. If labels in different choices are different, it means
-      //    that we can only print a base name.
-      if (name.isSpecial()) {
-        DE.diagnose(commonAnchor->getLoc(),
-                    diag::no_overloads_match_exactly_in_call_special,
-                    decl->getDescriptiveKind());
-      } else if (name.isOperator()) {
-        auto operatorId = name.getBaseIdentifier();
-        diagnoseOperatorAmbiguity(*this, operatorId, solutions,
-                                  commonCalleeLocator);
-      } else if (llvm::all_of(distinctChoices,
-                              [&name](const ValueDecl *choice) {
-                                return choice->getFullName() == name;
-                              })) {
-        DE.diagnose(commonAnchor->getLoc(),
-                    diag::no_overloads_match_exactly_in_call,
-                    decl->getDescriptiveKind(), name);
-      } else {
-        DE.diagnose(commonAnchor->getLoc(),
-                    diag::no_overloads_match_exactly_in_call_no_labels,
-                    decl->getDescriptiveKind(), name.getBaseName());
-      }
-    };
+    // Emit an error message for the ambiguity.
+    if (aggregatedFixes.size() == 1 &&
+        aggregatedFixes.front().first.first->isForContextualType()) {
+      auto *anchor = aggregatedFixes.front().first.first->getAnchor();
+      auto baseName = name.getBaseName();
+      DE.diagnose(commonAnchor->getLoc(), diag::no_candidates_match_result_type,
+                  baseName.userFacingName(), getContextualType(anchor));
+    } else if (name.isOperator()) {
+      diagnoseOperatorAmbiguity(*this, name.getBaseIdentifier(), solutions,
+                                commonCalleeLocator);
+      return true;
+    } else {
+      bool isApplication = llvm::find_if(ArgumentInfos, [&](const auto argInfo) {
```

Comment:
More of a question, since the value of `llvm::find_if` seems not to be used, just transformed into a bool with `!= ArgumentInfos.end()`, would be an option to use something `llvm::any_of` which will already return a bool?

Reply:
You're right, `llvm::any_of` would be better here. Thank you!

@xedin do you want to make this change or should I?

Reply:
I will fix it together with all my typos a bit later today :)

Reply:
Fixed!